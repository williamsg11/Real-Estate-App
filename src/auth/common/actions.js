export const ACTION_TYPES = {
  LOGIN_REQUEST: 'LOGIN_REQUEST',
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
  LOGIN_FAILURE: 'LOGIN_FAILURE',
  REGISTER_REQUEST: 'REGISTER_REQUEST',
  REGISTER_SUCCESS: 'REGISTER_SUCCESS',
  REGISTER_FAILURE: 'REGISTER_FAILURE',
  LOGOUT: 'LOGOUT',
  FORGOT_PASSWORD_REQUEST: 'FORGOT_PASSWORD_REQUEST',
  FORGOT_PASSWORD_SUCCESS: 'FORGOT_PASSWORD_SUCCESS',
  FORGOT_PASSWORD_FAILURE: 'FORGOT_PASSWORD_FAILURE',
  RECOVER_PASSWORD_REQUEST: 'RECOVER_PASSWORD_REQUEST',
  RECOVER_PASSWORD_SUCCESS: 'RECOVER_PASSWORD_SUCCESS',
  RECOVER_PASSWORD_FAILURE: 'RECOVER_PASSWORD_FAILURE',
  PASSWORD_UPDATE_REQUEST: 'PASSWORD_UPDATE_REQUEST',
  PASSWORD_UPDATE_SUCCESS: 'PASSWORD_UPDATE_SUCCESS',
  PASSWORD_UPDATE_FAILURE: 'PASSWORD_UPDATE_FAILURE',
};

function login(credentials) {
  return {
    type: ACTION_TYPES.LOGIN_REQUEST,
    credentials,
  };
}

function logout() {
  return {
    type: ACTION_TYPES.LOGOUT,
  };
}

function forgotPassword(params) {
  return {
    type: ACTION_TYPES.FORGOT_PASSWORD_REQUEST,
    params,
  };
}

function recoverPassword(params) {
  return {
    type: ACTION_TYPES.RECOVER_PASSWORD_REQUEST,
    params,
  };
}
function updatePassword(params) {
  return {
    type: ACTION_TYPES.PASSWORD_UPDATE_REQUEST,
    params,
  };
}

function register(params) {
  return {
    type: ACTION_TYPES.REGISTER_REQUEST,
    params,
  };
}

export const ACTIONS = {
  login,
  register,
  logout,
  recoverPassword,
  forgotPassword,
  updatePassword,
};
