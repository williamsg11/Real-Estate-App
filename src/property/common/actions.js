/**
 * @flow
 */
export const ACTION_TYPES = {
  PROPERTY_REQUEST: 'PROPERTY_REQUEST',
  PROPERTY_SUCCESS: 'PROPERTY_SUCCESS',
  PROPERTY_FAILURE: 'PROPERTY_FAILURE',
  PROPERTY_FAVORITE_REQUEST: 'PROPERTY_FAVORITE_REQUEST',
  PROPERTY_FAVORITE_SUCCESS: 'PROPERTY_FAVORITE_SUCCESS',
  PROPERTY_FAVORITE_FAILURE: 'PROPERTY_FAVORITE_FAILURE',
  PROPERTY_FAVORITE_OPTIMISTIC_UPDATE: 'PROPERTY_FAVORITE_OPTIMISTIC_UPDATE',
  PROPERTY_MAPVIEW_CHANGE: 'PROPERTY_MAPVIEW_CHANGE',
  PROPERTY_SET_FILTERS_REQUEST: 'PROPERTY_SET_FILTERS_REQUEST',
  PROPERTY_ADD_ITEM_TO_HISTORY: 'PROPERTY_ADD_ITEM_TO_HISTORY',
  PROPERTY_REMOVE_ITEM_FROM_HISTORY: 'PROPERTY_REMOVE_ITEM_FROM_HISTORY',
  PROPERTY_TYPE_CHANGE: 'PROPERTY_TYPE_CHANGE',
  PROPERTY_INCREMENT_VIEW_COUNT: 'PROPERTY_INCREMENT_VIEW_COUNT',
  PROPERTY_SAVE_REQUEST: 'PROPERTY_SAVE_REQUEST',
  PROPERTY_SAVE_SUCCESS: 'PROPERTY_SAVE_SUCCESS',
  PROPERTY_SAVE_FAILURE: 'PROPERTY_SAVE_FAILURE',
  PROPERTY_EDIT_REQUEST: 'PROPERTY_EDIT_REQUEST',
  PROPERTY_DELETE_REQUEST: 'PROPERTY_DELETE_REQUEST',
  PROPERTY_DELETE_SUCCESS: 'PROPERTY_DELETE_SUCCESS',
  PROPERTY_DELETE_FAILURE: 'PROPERTY_DELETE_FAILURE',
  PROPERTY_RELATED_REQUEST: 'PROPERTY_RELATED_REQUEST',
  PROPERTY_RELATED_SUCCESS: 'PROPERTY_RELATED_SUCCESS',
  PROPERTY_RELATED_FAILURE: 'PROPERTY_RELATED_FAILURE',
  PROPERTY_SEARCH_REQUEST: 'PROPERTY_SEARCH_REQUEST',
  PROPERTY_SEARCH_SUCCESS: 'PROPERTY_SEARCH_SUCCESS',
  PROPERTY_SEARCH_FAILURE: 'PROPERTY_SEARCH_FAILURE',
  PROPERTY_REQUESTED_URL: 'PROPERTY_REQUESTED_URL',
  PROPERTY_RESET_NEXT_PAGE_URL: 'PROPERTY_RESET_NEXT_PAGE_URL',
  PROPERTY_RESET: 'PROPERTY_RESET',
  MY_PROPERTY_REQUEST: 'MY_PROPERTY_REQUEST',
  MY_PROPERTY_SUCCESS: 'MY_PROPERTY_SUCCESS',
  MY_PROPERTY_FAILURE: 'MY_PROPERTY_FAILURE',
  FAVORITES_SUCCESS: 'FAVORITES_SUCCESS',
  FAVORITES_REQUEST: 'FAVORITES_REQUEST',
  FAVORITES_FAILURE: 'FAVORITES_FAILURE',
  FILTER_RESET: 'FILTER_RESET',
  FILTER_UPDATE: 'FILTER_UPDATE',
  LISTING_RESET: 'LISTING_RESET',
  LISTING_UPDATE_ITEM: 'LISTING_UPDATE_ITEM',
  EDIT_LISTING_UPDATE_ITEM: 'EDIT_LISTING_UPDATE_ITEM',
  PROPERTY_ADD_ITEM_TO_HISTORY_REQUEST: 'PROPERTY_ADD_ITEM_TO_HISTORY_REQUEST',
};

function fetchProperties(params) {
  return {
    type: ACTION_TYPES.PROPERTY_REQUEST,
    params,
  };
}

function fetchMyProperties(params) {
  return {
    type: ACTION_TYPES.MY_PROPERTY_REQUEST,
    params,
  };
}

function fetchFavorites(params) {
  return {
    type: ACTION_TYPES.FAVORITES_REQUEST,
    params,
  };
}

function updateFilter({propertyType, field, value}) {
  return {
    type: ACTION_TYPES.FILTER_UPDATE,
    propertyType,
    field,
    value,
  };
}

function resetProperty() {
  return {
    type: ACTION_TYPES.PROPERTY_RESET,
  };
}
function resetPropertyNextPageURL() {
  return {
    type: ACTION_TYPES.PROPERTY_RESET_NEXT_PAGE_URL,
  };
}

function favoriteProperty(property) {
  return {
    type: ACTION_TYPES.PROPERTY_FAVORITE_REQUEST,
    params: {
      itemID: property._id,
      newItemAttributes: {
        isFavorited: !property.isFavorited,
      },
    },
  };
}

function changeListingValue(payload) {
  return {
    type: ACTION_TYPES.LISTING_UPDATE_ITEM,
    payload,
  };
}
function changeEditingListingValue(payload) {
  return {
    type: ACTION_TYPES.EDIT_LISTING_UPDATE_ITEM,
    payload,
  };
}

function resetFilter(): void {
  return {
    type: ACTION_TYPES.FILTER_RESET,
  };
}

function saveProperty(payload: object): void {
  return {
    type: ACTION_TYPES.PROPERTY_SAVE_REQUEST,
    payload: payload,
  };
}
function deleteProperty(property: object): void {
  return {
    type: ACTION_TYPES.PROPERTY_DELETE_REQUEST,
    params: {
      itemID: property._id,
    },
  };
}

function editProperty(property) {
  return {
    type: ACTION_TYPES.PROPERTY_EDIT_REQUEST,
    payload: property,
  };
}

function changeMapView() {
  return {
    type: ACTION_TYPES.PROPERTY_MAPVIEW_CHANGE,
  };
}
function changePropertyType(propertyType) {
  return {
    type: ACTION_TYPES.PROPERTY_TYPE_CHANGE,
    propertyType: propertyType,
  };
}

function setFilters({propertyType, filters}) {
  return {
    type: ACTION_TYPES.PROPERTY_SET_FILTERS_REQUEST,
    filters,
    propertyType,
  };
}

function removeFromHistory(payload: object): void {
  return {
    type: ACTION_TYPES.PROPERTY_REMOVE_ITEM_FROM_HISTORY,
    payload: payload,
  };
}
function incrementViews(propertyID: number): void {
  return {
    type: ACTION_TYPES.PROPERTY_INCREMENT_VIEW_COUNT,
    propertyID,
  };
}

export const ACTIONS = {
  fetchProperties,
  fetchMyProperties,
  resetProperty,
  favoriteProperty,
  fetchFavorites,
  changeListingValue,
  changeEditingListingValue,
  saveProperty,
  resetFilter,
  changeMapView,
  setFilters,
  changePropertyType,
  incrementViews,
  removeFromHistory,
  deleteProperty,
  editProperty,
  resetPropertyNextPageURL,
  updateFilter,
};
