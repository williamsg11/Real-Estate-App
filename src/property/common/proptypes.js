/* @flow */

import {PropTypes} from 'react';

export const CountryPropType = PropTypes.shape({
  fullName: PropTypes.string.isRequired,
  abbr: PropTypes.string.isRequired,
  currency: PropTypes.string.isRequired,
  coords: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
  }),
});

// export const NavigationStatePropType = PropTypes.shape({
//   routes: PropTypes.arrayOf(NavigationRoutePropType).isRequired,
//   index: PropTypes.number.isRequired,
// });

// export const SceneRendererPropType = {
//   layout: PropTypes.shape({
//     measured: PropTypes.bool.isRequired,
//     height: PropTypes.number.isRequired,
//     width: PropTypes.number.isRequired,
//   }).isRequired,
//   navigationState: NavigationStatePropType.isRequired,
//   position: PropTypes.instanceOf(Animated.Value).isRequired,
//   jumpToIndex: PropTypes.func.isRequired,
//   getLastPosition: PropTypes.func.isRequired,
//   subscribe: PropTypes.func.isRequired,
// };
