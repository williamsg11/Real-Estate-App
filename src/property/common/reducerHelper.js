let salePrice1 = [
  'Any',
  '10K',
  '25K',
  '50K',
  '100K',
  '200K',
  '300K',
  '400K',
  '500K',
  '1M',
  '5M',
  '10M',
];

let salePrice2 = [
  'Any',
  '25K',
  '50K',
  '100K',
  '250K',
  '500K',
  '1M',
  '5M',
  '10M',
  '25M',
  '50M',
  '100M',
];

let rentPrice1 = [
  'Any',
  '200',
  '300',
  '400',
  '500',
  '600',
  '700',
  '800',
  '900',
  '1000',
  '2000',
  '5000',
];
let rentPrice2 = [
  'Any',
  '1500',
  '3000',
  '4000',
  '5000',
  '6000',
  '7000',
  '8000',
  '9000',
  '10K',
  '15K',
  '20K',
];

let sharePrice1 = [
  'Any',
  '100',
  '150',
  '200',
  '250',
  '300',
  '350',
  '400',
  '450',
  '500',
  '1000',
  '2500',
  '5000',
];
let sharePrice2 = [
  'Any',
  '500',
  '750',
  '1000',
  '1500',
  '2000',
  '3000',
  '4000',
  '5000',
  '10K',
  '15K',
  '20K',
];

let countries = ['KW', 'OM', 'BH', 'SA', 'AE', 'QA'];
let [KW, OM, BH, SA, AE, QA] = countries;

let propertyTypes = ['For Sale', 'For Rent', 'For Share'];
let [ForSale, ForRent, ForShare] = propertyTypes;

let sortOptions = ['New', 'Price(Low-High)', 'Price(High-Low)', 'Old'];

let filters = {
  [ForSale]: {
    priceFrom: 'Any',
    priceTo: 'Any',
    parking: 'Any',
    bedroom: 'Any',
    bathroom: 'Any',
    category: 'Any',
    sortBy: sortOptions[0],
    searchString: '',
  },
  [ForRent]: {
    priceFrom: 'Any',
    priceTo: 'Any',
    parking: 'Any',
    bedroom: 'Any',
    bathroom: 'Any',
    category: 'Any',
    sortBy: sortOptions[0],
    searchString: '',
  },
  [ForShare]: {
    priceFrom: 'Any',
    priceTo: 'Any',
    parking: 'Any',
    bedroom: 'Any',
    bathroom: 'Any',
    category: 'Any',
    sortBy: sortOptions[0],
    searchString: '',
  },
};

let priceForRent = {
  [KW]: rentPrice1,
  [OM]: rentPrice1,
  [BH]: rentPrice1,
  [SA]: rentPrice2,
  [AE]: rentPrice2,
  [QA]: rentPrice2,
};
let priceForShare = {
  [KW]: sharePrice1,
  [OM]: sharePrice1,
  [BH]: sharePrice1,
  [SA]: sharePrice2,
  [AE]: sharePrice2,
  [QA]: sharePrice2,
};

let priceForSale = {
  [KW]: salePrice1,
  [OM]: salePrice1,
  [BH]: salePrice1,
  [SA]: salePrice2,
  [AE]: salePrice2,
  [QA]: salePrice2,
};

let categoriesArr = {
  [ForSale]: [
    'Villa',
    'House',
    'Apartment',
    'Chalet',
    'Land & Farm',
    'Store',
    'Storage',
    'Office',
  ],
  [ForRent]: [
    'Villa',
    'House',
    'Apartment',
    'Chalet',
    'Land & Farm',
    'Store',
    'Storage',
    'Office',
  ],
  [ForShare]: [
    'Villa',
    'House',
    'Apartment',
    'Chalet',
    'Land & Farm',
    'Store',
    'Storage',
    'Office',
  ],
};

let pricesArr = {
  [ForSale]: priceForSale,
  [ForRent]: priceForRent,
  [ForShare]: priceForShare,
};

let genders = ['Any', 'Family', 'Bachelor', 'Bachelorette'];

let addMetas = {
  bedroomsArr: ['Studio', '1', '2', '3', '4', '5', '6', '7', '7+'],
  bathroomsArr: ['1', '2', '3', '4', '5', '6', '7', '7+'],
  parkingArr: ['N/A', '1', '2', '3', '4', '4+'],
};

let searchMetas = {
  bedroomsArr: ['Any', 'Studio', '1', '1+', '2', '2+', '3', '3+', '4', '4+'],
  bathroomsArr: ['Any', '1', '1+', '2', '2+', '3', '3+', '4', '4+'],
  parkingArr: ['Any', '1', '1+', '2', '2+', '3', '3+', '4', '4+'],
};

let amenities = [
  'Swimming Pool',
  'Playground',
  'Gym',
  'Furnished',
  'Furnished Kitchen',
  'Central AC',
  'Balcony',
  'Lift',
];

let nearByPlaces = [
  'Mosque',
  'Hotels',
  'Gas Station',
  'Co-op',
  'Entertainment Park',
  'Beach',
  'Water Tank',
  'Malls',
  'Cinema',
  'Restaurants',
  'Schools',
];

let attributes = {
  type: '',
  category: '',
  address: {
    city: 'Kuwait City',
    state: 'Kuwait',
    country: 'Kuwait',
    latitude: 29.3667,
    longitude: 47.9667,
  },
  meta: {
    description: '',
    price: '',
    area: '',
    bedroom: 1,
    bathroom: 1,
    parking: 'N/A',
    gender: '',
    email: '',
    phone1: '',
    phone2: '',
  },
  images: [],
  amenities: [],
  nearByPlaces: [],
};

module.exports = {
  filters,
  amenities,
  nearByPlaces,
  attributes,
  propertyTypes,
  genders,
  searchMetas,
  addMetas,
  sortOptions,
  categories: categoriesArr,
  prices: pricesArr,
};
