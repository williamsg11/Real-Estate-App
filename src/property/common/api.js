import {API_URL} from '../../env';
import {fetchAPI} from '../../common/api';

function fetchProperties(params) {
  const url = `${API_URL}/properties${params}`;
  return fetchAPI(url);
}

function fetchMyProperties(params) {
  const url = `${API_URL}/properties/self${params}`;
  return fetchAPI(url);
}

function fetchFavorites(params) {
  const url = `${API_URL}/favorites${params}`;
  return fetchAPI(url);
}

function favoriteProperty(urlParams, body) {
  const url = `${API_URL}/favorites${urlParams}`;
  return fetchAPI(url, 'POST', body);
}

function deleteProperty(urlParams, body) {
  const url = `${API_URL}/properties/delete${urlParams}`;
  return fetchAPI(url, 'POST', body);
}

function saveProperty(body, urlParams) {
  const url = `${API_URL}/properties?${urlParams}`;
  return fetchAPI(url, 'POST', body);
}

function uploadImage(propertyID, body) {
  const url = `${API_URL}/properties/${propertyID}/images/upload`;
  return fetchAPI(url, 'POST', body, true);
}

function incrementViews(propertyID) {
  const url = `${API_URL}/properties/${propertyID}/increment-view`;
  try {
    fetchAPI(url);
  } catch (e) {
    console.log('e', e);
  }
}

export const API = {
  fetchProperties,
  fetchMyProperties,
  fetchFavorites,
  favoriteProperty,
  saveProperty,
  uploadImage,
  incrementViews,
  deleteProperty,
};
