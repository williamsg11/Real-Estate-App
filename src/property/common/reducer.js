import {ACTION_TYPES} from './actions';
import merge from 'lodash/merge';
import isArray from 'lodash/isArray';
import concat from 'lodash/concat';
import map from 'lodash/map';
import union from 'lodash/union';
import isEqual from 'lodash/isEqual';
import {
  filters,
  amenities,
  nearByPlaces,
  attributes,
  propertyTypes,
  genders,
  addMetas,
  searchMetas,
  sortOptions,
  categories,
  prices,
} from './reducerHelper';

const initialState = {
  isFetching: false,
  isFavoritesFetching: false,
  isRelatedFetching: false,
  isMyPropertiesFetching: false,
  isSaving: false,
  error: null,
  nextPageUrl: undefined,
  nextPageFavoritesUrl: undefined,
  mapView: false,
  showRelated: false,
  results: [],
  relatedResults: [],
  addListing: {
    stage: 1,
    attributes: {
      ...attributes,
    },
  },
  editListing: {
    stage: 1,
    attributes: {
      ...attributes,
    },
  },
};

export default function propertyReducer(state = initialState, action = {}) {
  switch (action.type) {
    case ACTION_TYPES.PROPERTY_REQUEST:
      return {
        ...state,
        isFetching: true,
        error: null,
        showRelated: false,
      };
    case ACTION_TYPES.PROPERTY_SUCCESS: {
      let results = [];
      const propertyCollections = action.payload.data;
      map(propertyCollections, entity => {
        results.push(entity._id);
      });
      return {
        ...state,
        isFetching: false,
        error: null,
        results: union(state.results, results),
        nextPageUrl: action.payload.next_page_url,
      };
    }
    case ACTION_TYPES.PROPERTY_FAILURE:
      return {...state, isFetching: false, error: action.error};

    case ACTION_TYPES.MY_PROPERTY_REQUEST:
      return {
        ...state,
        isMyPropertiesFetching: true,
        error: null,
      };
    case ACTION_TYPES.MY_PROPERTY_SUCCESS: {
      let myPropertyResults = [];
      let propertyCollections = action.payload.data;
      map(propertyCollections, entity => {
        myPropertyResults.push(entity._id);
      });
      return {
        ...state,
        isMyPropertiesFetching: false,
        error: null,
        results: union(state.results, myPropertyResults),
        myNextPageUrl: action.payload.next_page_url,
      };
    }
    case ACTION_TYPES.MY_PROPERTY_FAILURE:
      return {...state, isMyPropertiesFetching: false, error: action.error};

    case ACTION_TYPES.FAVORITES_REQUEST:
      return {
        ...state,
        isFavoritesFetching: true,
        error: null,
      };
    case ACTION_TYPES.FAVORITES_SUCCESS:
      const favoriteResults = [];
      const favoritePropertyCollections = action.payload.data;
      map(favoritePropertyCollections, entity => {
        favoriteResults.push(entity._id);
      });
      return {
        ...state,
        nextPageFavoritesUrl: action.payload.next_page_url,
        results: union(state.results, favoriteResults),
        isFavoritesFetching: false,
      };
    case ACTION_TYPES.FAVORITES_FAILURE:
      return {...state, isFavoritesFetching: false, error: action.error};

    case ACTION_TYPES.PROPERTY_RELATED_REQUEST:
      return {
        ...state,
        isRelatedFetching: true,
        error: null,
      };
    case ACTION_TYPES.PROPERTY_RELATED_SUCCESS:
      let results = [];
      const propertyCollections = action.payload.data;
      map(propertyCollections, entity => {
        results.push(entity._id);
      });
      return {
        ...state,
        isRelatedFetching: false,
        error: null,
        showRelated: true,
        relatedResults: union(state.relatedResults, results),
      };
    case ACTION_TYPES.PROPERTY_RELATED_FAILURE:
      return {
        ...state,
        isRelatedFetching: false,
        error: action.error,
      };

    case ACTION_TYPES.PROPERTY_SAVE_REQUEST:
      return {
        ...state,
        isSaving: true,
      };
    case ACTION_TYPES.PROPERTY_SAVE_SUCCESS:
      return {
        ...state,
        isSaving: false,
        addListing: {
          ...state.addListing,
          stage: 1,
          attributes: {
            ...initialState.addListing.attributes,
          },
        },
        editListing: {
          ...state.editListing,
          stage: 1,
          attributes: {
            ...initialState.editListing.attributes,
          },
        },
      };
    case ACTION_TYPES.PROPERTY_SAVE_FAILURE:
      return {
        ...state,
        isSaving: false,
      };

    case ACTION_TYPES.PROPERTY_RESET:
      return {
        ...state,
        results: [],
        nextPageUrl: undefined,
        showRelated: false,
      };
    case ACTION_TYPES.PROPERTY_RESET_NEXT_PAGE_URL:
      return {
        ...state,
        nextPageUrl: undefined,
      };
    case ACTION_TYPES.PROPERTY_MAPVIEW_CHANGE:
      return {
        ...state,
        mapView: !state.mapView,
      };

    case ACTION_TYPES.LISTING_RESET:
      return {
        ...state,
        addListing: {
          ...state.addListing,
        },
      };
    case ACTION_TYPES.LISTING_UPDATE_ITEM:
      if (action.payload.replace) {
        const {key, item} = action.payload;
        const oldState = state.addListing.attributes[key];
        let newState;

        if (isArray(item)) {
          newState = union(oldState, item);
        } else {
          if (oldState.includes(item)) {
            // remove item
            newState = oldState.filter(value => value !== item);
          } else {
            newState = union(oldState, [item]);
          }
        }
        return {
          ...state,
          addListing: {
            ...state.addListing,
            attributes: {
              ...state.addListing.attributes,
              [key]: newState,
            },
          },
        };
      }
      return {
        ...state,
        addListing: merge({}, state.addListing, action.payload),
      };

    case ACTION_TYPES.EDIT_LISTING_UPDATE_ITEM:
      if (action.payload.replace) {
        const {key, item} = action.payload;
        const oldState = state.editListing.attributes[key];
        let newState;

        if (isArray(item)) {
          newState = union(oldState, item);
        } else {
          if (oldState.includes(item)) {
            // remove item
            newState = oldState.filter(value => value !== item);
          } else {
            newState = union(oldState, [item]);
          }
        }
        return {
          ...state,
          editListing: {
            ...state.editListing,
            attributes: {
              ...state.editListing.attributes,
              [key]: newState,
            },
          },
        };
      }
      return {
        ...state,
        editListing: merge({}, state.editListing, action.payload),
      };

    case ACTION_TYPES.PROPERTY_EDIT_REQUEST:
      return {
        ...state,
        editListing: {
          ...state.editListing,
          attributes: action.payload,
        },
      };

    case ACTION_TYPES.PROPERTY_DELETE_REQUEST:
      return {
        ...state,
        results: state.results.filter(
          itemID => itemID !== action.params.itemID,
        ),
      };

    default:
      return state;
  }
}

export function propertyHistoryReducer(state = [], action = {}) {
  switch (action.type) {
    case ACTION_TYPES.PROPERTY_ADD_ITEM_TO_HISTORY:
      return concat(state, action.payload);
      break;
    case ACTION_TYPES.PROPERTY_REMOVE_ITEM_FROM_HISTORY:
      return state.filter(history => {
        return !isEqual(history, action.payload);
      });
    default:
      return state;
  }
}

let initialFilters = {
  propertyTypes,
  selectedPropertyType: propertyTypes[0],
  filters,
  amenities,
  nearByPlaces,
  genders,
  addMetas,
  searchMetas,
  sortOptions,
  categories,
  prices,
};

export function propertyOptionsReducer(state = initialFilters, action = {}) {
  switch (action.type) {
    case ACTION_TYPES.PROPERTY_SET_FILTERS_REQUEST:
      return {
        ...state,
        selectedPropertyType: action.propertyType,
        filters: {
          ...state.filters,
          [action.propertyType]: {
            ...state.filters[action.propertyType],
            ...action.filters,
          },
        },
      };
      break;

    case ACTION_TYPES.PROPERTY_TYPE_CHANGE:
      return {
        ...state,
        selectedPropertyType: action.propertyType,
      };
    case ACTION_TYPES.FILTER_UPDATE:
      return {
        ...state,
        filters: {
          ...state.filters,
          [action.propertyType]: {
            ...state.filters[action.propertyType],
            [action.field]: action.value,
          },
        },
      };
    case ACTION_TYPES.FILTER_RESET:
      return {
        ...state,
        filters: state.filters,
      };
    default:
      return state;
  }
}
