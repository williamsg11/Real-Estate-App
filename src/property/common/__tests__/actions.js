import {ACTIONS} from '../../common/actions';

describe('App Component Actions', () => {
  test('fetchProperties action', () => {
    expect(
      ACTIONS.fetchProperties({
        id: 1,
      }),
    ).toMatchSnapshot();
  });

  test('fetchFavorites action', () => {
    expect(
      ACTIONS.fetchFavorites({
        id: 1,
      }),
    ).toMatchSnapshot();
  });

  test('updateFilter action', () => {
    expect(ACTIONS.updateFilter('bedroom', 1)).toMatchSnapshot();
  });

  test('reset property action', () => {
    expect(ACTIONS.resetProperty()).toMatchSnapshot();
  });

  test('changeListingValue action', () => {
    expect(
      ACTIONS.changeListingValue('attributes', {
        title: 'property title',
        address: {
          city: 'Kuwait City',
          state: 'Kuwait City',
          country: 'Kuwait',
          latitude: 29.3667,
          longitude: 47.9667,
        },
        meta: {
          bedroom: 'Studio',
          bathroom: '1',
          kitchen: '1',
          area: '220.5',
          parking: '1',
        },
        images: [],
        tags: ['New', 'Duplex'],
        amenities: ['Swimming Pool'],
      }),
    ).toMatchSnapshot();
  });

  test('reset filter', () => {
    expect(ACTIONS.resetFilter()).toMatchSnapshot();
  });

  test('reset filter', () => {
    expect(ACTIONS.saveProperty()).toMatchSnapshot();
  });

  test('add to history', () => {
    expect(
      ACTIONS.setFilters({'For Sale': {country: 'Kuwait'}}),
    ).toMatchSnapshot();
  });

  test('remove from history', () => {
    expect(
      ACTIONS.removeFromHistory({'For Sale': {country: 'Kuwait'}}),
    ).toMatchSnapshot();
  });

  test('edit property', () => {
    expect(
      ACTIONS.editProperty({_id: '123', title: 'property'}),
    ).toMatchSnapshot();
  });

  test('delete property', () => {
    expect(
      ACTIONS.deleteProperty({_id: '123', title: 'property'}),
    ).toMatchSnapshot();
  });
});
