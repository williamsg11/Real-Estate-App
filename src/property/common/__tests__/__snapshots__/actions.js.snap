// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Actions add to history 1`] = `
Object {
  "filters": undefined,
  "propertyType": undefined,
  "type": "PROPERTY_SET_FILTERS_REQUEST",
}
`;

exports[`App Component Actions changeListingValue action 1`] = `
Object {
  "payload": "attributes",
  "type": "LISTING_UPDATE_ITEM",
}
`;

exports[`App Component Actions delete property 1`] = `
Object {
  "params": Object {
    "itemID": "123",
  },
  "type": "PROPERTY_DELETE_REQUEST",
}
`;

exports[`App Component Actions edit property 1`] = `
Object {
  "payload": Object {
    "_id": "123",
    "title": "property",
  },
  "type": "PROPERTY_EDIT_REQUEST",
}
`;

exports[`App Component Actions fetchFavorites action 1`] = `
Object {
  "params": Object {
    "id": 1,
  },
  "type": "FAVORITES_REQUEST",
}
`;

exports[`App Component Actions fetchProperties action 1`] = `
Object {
  "params": Object {
    "id": 1,
  },
  "type": "PROPERTY_REQUEST",
}
`;

exports[`App Component Actions remove from history 1`] = `
Object {
  "payload": Object {
    "For Sale": Object {
      "country": "Kuwait",
    },
  },
  "type": "PROPERTY_REMOVE_ITEM_FROM_HISTORY",
}
`;

exports[`App Component Actions reset filter 1`] = `
Object {
  "type": "FILTER_RESET",
}
`;

exports[`App Component Actions reset filter 2`] = `
Object {
  "payload": undefined,
  "type": "PROPERTY_SAVE_REQUEST",
}
`;

exports[`App Component Actions reset property action 1`] = `
Object {
  "type": "PROPERTY_RESET",
}
`;

exports[`App Component Actions updateFilter action 1`] = `
Object {
  "field": undefined,
  "propertyType": undefined,
  "type": "FILTER_UPDATE",
  "value": undefined,
}
`;
